gamemode/modules/commands/sv_commands.lua

```diff
@@ -0,0 +1,260 @@
+    local str = ""
+
+    local insideQuotes = false
+
+    local i = 1
+
+    while i <= string.len( argsStr ) do
+        local char = string.sub( argsStr, i, i )
+
+        if char == "\"" then
+            insideQuotes = not insideQuotes
+        elseif char == "\\" then
+            i = i + 1
+
+            char = string.sub( argsStr, i, i )
+            str = str .. char
+        elseif char == " " and not insideQuotes then
+            table.insert( args, str )
+            str = ""
+        else
+            str = str .. char
+        end
+
+        i = i + 1
+    end
+
+    if str ~= "" then
+        table.insert( args, str )
+    end
+
+    return args
+end
+
+local function processCommand( ply, command, args )
+    local commandInfo = commands.list[command]
+
+    if commandInfo then
+        if not RVR.isUserGroup( ply, commandInfo.userGroup ) then
+            return "You need to be " .. RVR.getGroupName( commandInfo.userGroup ) .. " to use this command"
+        end
+
+        local newArgs, errorMsg = commands.checkArguments( commandInfo.argNames, commandInfo.argTypes, args )
+
+        if errorMsg then
+            return errorMsg, true
+        end
+
+        local msg = commandInfo.func( ply, unpack( newArgs ) )
+
+        return msg, true
+    end
+end
+
+function commands.register( names, argNames, argTypes, userGroup, func, desc )
+    if #argNames ~= #argTypes then
+        error( "There must be the same amount of argument names and types" )
+    end
+
+    if type( names ) == "string" then
+        names = { names }
+    end
+
+    for _, name in ipairs( names ) do
+        if commands.list[name] then
+            error( "A command with the same name (" .. name .. ") already exists" )
+        end
+
+        for _, argType in ipairs( argTypes ) do
+            if not commands.types[argType] then
+                error( "\"" .. argType .. "\" is not a valid argument type" )
+            end
+        end
+
+        local description = "Usage: " .. name
+
+        for i, argName in ipairs( argNames ) do
+            local argType = argTypes[i]
+
+            description = description .. " " .. argName .. ":" .. argType
+        end
+
+        if desc then
+            description = description .. "\nDescription: " .. desc
+        end
+
+        commands.list[name] = {
+            argNames = argNames,
+            argTypes = argTypes,
+            userGroup = userGroup,
+            func = func,
+            description = description
+        }
+    end
+end
+
+local function onPlayerSay( ply, text )
+    if string.sub( text, 1, 1 ) ~= "!" then return end
+
+    text = string.sub( text, 2 )
+
+    local args = processArguments( text )
+    local command = table.remove( args, 1 )
+
+    local msg, validCommand = processCommand( ply, command, args )
+
+    if msg then
+        ply:ChatPrint( msg )
+    end
+
+    if not validCommand then
+        ply:ChatPrint( "Command \"" .. command .. "\" does not exist" )
```

Comment:
Probably still want to `return ""` even if the command is invalid. Would be annoying if a spelling mistake gave away something you were trying to do privately.